### PR TITLE
Store test file in S3 as well for every TestSuite

### PR DESCRIPTION
**Pull request by janeyx99**
We want to store the file names that triggers each test suite so that we can use this data for categorizing those test files.

~~After considering several solutions, this one is the most backwards compatible, and the current test cases in test_testing.py for print test stats don't break.~~

The previous plan did not work, as there are multiple Python test jobs that spawn the same suites. Instead, the new S3 format will store test files (e.g., `test_nn` and `distributed/test_distributed_fork`) which will contain the suites they spawn, which will contain the test cases run within the suite. (Currently, there is no top layer of test files.)

Because of this major structural change, a lot of changes have now been made (thank you @samestep!) to test_history.py and print_test_stats.py to make this new format backwards compatible.

Old test plan:
Make sure that the data is as expected in S3 after https://github.com/pytorch/pytorch/pull/52873 finishes.

Test plan:
Added tests to test_testing.py which pass, and CI.

### CODE REVIEW

File: torch/testing/_internal/print_test_stats.py [hunk 1]
```diff
@@ -13,7 +14,7 @@
 from pathlib import Path
 from typing import (Any, DefaultDict, Dict, Iterable, Iterator, List, Optional,
                     Tuple)
-from xml.dom import minidom  # type: ignore[import]
```
Comment (samestep) [inline torch/testing/_internal/print_test_stats.py:16]: Why did you remove this comment? I thought it was necessary for `mypy` to pass?

Comment (janeyx99) [reply torch/testing/_internal/print_test_stats.py:16]: I don't think it is?

Comment (samestep) [reply torch/testing/_internal/print_test_stats.py:16]: Ok cool 😃 that's nice if it's not required

**Review by samestep — APPROVED** [review]

File: torch/testing/_internal/print_test_stats.py (hunk 1 above)
Comment (janeyx99) [reply torch/testing/_internal/print_test_stats.py:16]: never mind...lol weird because locally it recommended I delete the comment:
image

Comment (samestep) [reply torch/testing/_internal/print_test_stats.py:16]: 😮 hmm that's... not good haha, I think I've run into a similar issue before but I forget exactly how to deal with it

File: tools/test_history.py [hunk 2]
```diff
@@ -44,7 +44,12 @@ class Case(TypedDict):
     skipped: bool
 
 
-class Suite(TypedDict):
+# The original format of the s3 reports did not have this filename field. This maintains backwards-compatibility
+class SuiteMeta(TypedDict, total=False):
+    filename: str
```
Comment (malfet) [inline tools/test_history.py:49]: `filename` would not be available for all the historical data, is it?
```suggestion
    filename: Optional[str]
```

**Review by malfet — APPROVED** [review]

File: tools/test_history.py (hunk 2 above)
Comment (walterddr) [reply tools/test_history.py:49]: I think Optional[str] annotation meant `filename` can be either a string or None. but still it requires to be explicitly set (e.g. `suiteMeta.filename = None` is still needed otherwise error will thrown). 
based on https://mypy.readthedocs.io/en/latest/more_types.html#totality, seems like the `total=False` part makes the key filename optional

**Review by walterddr — APPROVED** [review]

**Comment by janeyx99** [comment]
Two concerns: @walterddr @malfet @samestep 
1. Some suitenames are potentially duplicated, e.g. TestDataloader in test_utils.py and test_dataloader.py. The fix to this seems easy enough--I renamed one of them to TestDataloaderUtils. However, is this the best fix we can have?
2. I intend to use this method of storing filenames to currently shard python-only tests. However, we get reports from cpp tests running the same suites, which messes up the script [log](https://app.circleci.com/pipelines/github/pytorch/pytorch/278242/workflows/8918c14a-1192-43c4-b429-1a1bd7f69a1c/jobs/11167419). There are two solutions here:
    a. Ignore non-Python tests when reporting test stats. This has the downside of **no longer storing data for non-Python** tests on S3, for example the libtorch tests and test_jit executables we run, though currently, that data seems to be mixed with the Python stats under the same suite. (Currently what this PR will try to do.)
    b. Stop being backwards compatible and store suites under filenames. This would not run into the issue of duplicate Suites with different filenames, and I'd change the current test report paths to contain the cpp executable that spawns these tests. This solution would require **BIG** changes to the current print_test_stats and test_history scripts, however.

File: torch/testing/_internal/print_test_stats.py [hunk 3]
```diff
@@ -24,6 +25,126 @@
 except ImportError:
     HAVE_BOTO3 = False
 
+
+# TODO: this is ugly duplicated code from run_test.py we should probably abstract out to common_utils.py
```
Comment (janeyx99) [inline torch/testing/_internal/print_test_stats.py:29]: Would common_utils be a good place for this, if we were to go with this solution?

Comment (samestep) [reply torch/testing/_internal/print_test_stats.py:29]: I'd say no, since we're planning on moving `print_test_stats.py` to `tools`; not sure how to solve the duplication though.

**Review by walterddr — APPROVED** [review]
looks good overall. thanks for updating the approach
